proposals/NNNN-hashable-enhancements.md

```diff
@@ -0,0 +1,925 @@
+* Previous Revision: [1](https://github.com/apple/swift-evolution/blob/...commit-ID.../proposals/NNNN-filename.md)
+* Previous Proposal: [SE-XXXX](XXXX-filename.md)
+-->
+
+## Table of contents
+
+* [Introduction](#intro)
+* [Motivation](#why)
+    - [The Status Quo](#status-quo)
+    - [Universal hash functions](#universal-hashing)
+* [Proposed solution](#proposed-solution)
+    - [The `Hasher` struct](#hasher)
+    - [The `hash(into:)` requirement](#hash-into)
+* [Detailed design](#detailed-design)
+    - [`Hasher`](#hasher-details)
+    - [`Hashable`](#hashable-details)
+* [Source compatibility](#source-compatibility)
+* [Effect on ABI stability](#abi)
+* [Effect on API resilience](#resilience)
+* [Alternatives considered](#alternatives)
+    - [Leaving `Hashable` as is](#leave-hashable-alone)
+    - [Defining a protocol refinement](#protocol-refinement)
+    - [Making `hash(into:)` generic over a `Hasher` protocol](#generic-hasher)
+    - [Change `hash(into:)` to take a closure instead of a new type](#closure-hasher)
+
+## <a name="intro">Introduction</a>
+
+This proposal introduces a new `Hasher` type representing the standard
+library's universal hash function, and it extends the `Hashable`
+protocol with a new `hash(into:)` requirement that expresses hashing
+in terms of `Hasher`. This new requirement is intended to replace the
+old `hashValue` property, which is deprecated.
+
+Switching to `hash(into:)` moves the choice of a hash function out of
+`Hashable` implementations, and into the standard library. This makes
+it considerably easier to manually conform types to `Hashable` -- the
+task of writing a custom implementation reduces to identifying the
+parts of a type that should contribute to the hash value.
+
+Standardizing on a single, high-quality hash function greatly improves
+the reliability of `Set` and `Dictionary`. The hash function can be
+specially selected and tuned to work well with the hash tables used by
+these collections, preventing hash collision patterns that would break
+the expected performance of common operations.
+
+`Hasher` is a resilient struct, enabling future versions of the
+standard library to improve the hash function without the need to
+change (or even recompile) existing code that implements `Hashable`.
+Not baking any particular hash function into `Hashable` types is
+especially important in case a weakness is found in the current
+algorithm.
+
+Swift-evolution thread: [Combining Hashes](https://forums.swift.org/t/combining-hashes/9082)
+
+[SE-0185]: 0185-synthesize-equatable-hashable.md
+[SE-0143]: 0143-conditional-conformances.md
+
+## <a name="why">Motivation</a>
+
+The Swift Standard Library includes two general-purpose hashing
+collections, `Set` and `Dictionary`. These collections are built
+around hash tables, which are probabilistic data structures; their
+expected performance critically depends on the expected distribution
```

Comment:
No, hash tables are not probabilistic data structures. Their performance profile might be probabilistic, but they always give the right answer. Contrast with https://en.wikipedia.org/wiki/Bloom_filter